### PR TITLE
[QoL] Settings Option: Shop Cursor Target

### PR DESCRIPTION
**Pull request by torranx**
## What are the changes the user will see?
New option under display: Shop Cursor Target - sets the cursor in shop phase based on the set option (defaults to Check Team)

## Why am I making these changes?

[highly requested QoL](https://discord.com/channels/1125469663833370665/1235481427924095016), previous author is MIA - cherry-picked his commits 

## What are the changes from a developer perspective?

no functional change, just a new option

### Screenshots/Videos
Screenshot 2024-08-21 at 1 47 08 AM

https://github.com/user-attachments/assets/f08de990-2d84-4cf8-9841-16bceed265b0

## How to test the changes?
get to the shop phase, set option and reroll

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [ ] Have I considered writing automated tests for the issue?
- [x] If I have text, did I add placeholders for them in locales?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/locales/fr/settings.ts [hunk 1]
```diff
@@ -98,5 +98,9 @@ export const settings: SimpleTranslationEntries = {
   "gamepadSupport": "Gamepad Support",
   "showBgmBar": "Titre de la musique",
   "moveTouchControls": "Déplacer les contrôles tactiles",
-  "shopOverlayOpacity": "Opacité boutique"
+  "shopOverlayOpacity": "Opacité boutique",
+  "shopRerollTarget": "Shop Post-Reroll Target",
+  "items": "Items",
+  "reroll": "Reroll",
+  "shop": "Shop"
```
Comment (Adri1) [inline src/locales/fr/settings.ts:105]:
```suggestion
  "shopRerollTarget": "Choix après relance",
  "items": "Obj. gratuits",
  "reroll": "Relance",
  "shop": "Obj. payants"
```

**Review by Adri1 — COMMENTED** [review]
French Added (to be tested in context)

**Comment by flx-sta** [comment]
Other settings have an enum for every option. It might make sense here too.

```ts
/**
 * Determines the cursor target after selecting the re-roll button
 */
export enum ShopPostRerollTarget {
   /** Cursor stays on re-roll button */ 
   REROLL,
   /** Cursor moves to the first item after re-roll */
    ITEMS,
    /** Cursor moves to the first shop item after re-roll */
    SHOP,
}

File: src/locales/de/settings.ts [hunk 2]
```diff
@@ -99,4 +99,8 @@ export const settings: SimpleTranslationEntries = {
   "showBgmBar": "Musiknamen anzeigen",
   "moveTouchControls": "Bewegung Touch Steuerung",
   "shopOverlayOpacity": "Shop Overlay Deckkraft",
+  "shopRerollTarget": "Shop Post-Reroll Target",
+  "items": "Items",
+  "reroll": "Reroll",
+  "shop": "Shop"
```
Comment (CodeTappert) [inline src/locales/de/settings.ts:105]:
```suggestion
  "shopRerollTarget": "Shop Ziel nach Neurollen",
  "items": "Items",
  "reroll": "Neurollen",
  "shop": "Shop"
```

**Comment by torranx** [comment]
Placing this back to draft, please pause on translations until PR is ready. I will also ask an updated translation from you guys @Adri1 @CodeTappert when it's ready.

File: src/locales/ko/settings.ts [hunk 3]
```diff
@@ -98,5 +98,9 @@ export const settings: SimpleTranslationEntries = {
   "gamepadSupport": "게임패드 지원",
   "showBgmBar": "BGM 제목 보여주기",
   "moveTouchControls": "터치 컨트롤 이동",
-  "shopOverlayOpacity": "상점 오버레이 투명도"
+  "shopOverlayOpacity": "상점 오버레이 투명도",
+  "shopRerollTarget": "Shop Post-Reroll Target",
+  "items": "Items",
+  "reroll": "Reroll",
+  "shop": "Shop"
```
Comment (sodaMelon) [inline src/locales/ko/settings.ts:105]:
```suggestion
  "shopRerollTarget": "상점 갱신-후 타겟팅",
  "items": "아이템",
  "reroll": "갱신 버튼",
  "shop": "상점"
```

**Review by sodaMelon — COMMENTED** [review]
added korean localization

File: src/locales/fr/settings.ts (hunk 1 above)
Comment (torranx) [reply src/locales/fr/settings.ts]: I've updated the keys, please suggest a new change. Thanks!

File: src/locales/de/settings.ts (hunk 2 above)
Comment (torranx) [reply src/locales/de/settings.ts]: I've updated the keys, please suggest a new change. Thanks!

File: src/locales/ko/settings.ts (hunk 3 above)
Comment (torranx) [reply src/locales/ko/settings.ts]: I've updated the keys, please suggest a new change. Thanks!

File: src/system/settings/settings.ts [hunk 4]
```diff
@@ -9,6 +9,7 @@ import { EaseType } from "#enums/ease-type";
 import { MoneyFormat } from "#enums/money-format";
 import { PlayerGender } from "#enums/player-gender";
 import { getIsInitialized, initI18n } from "#app/plugins/i18n.js";
+import { ShopCursorTarget } from "#app/enums/shop-cursor-target.js";
```
Comment (flx-sta) [inline src/system/settings/settings.ts:12]:
```suggestion
import { ShopCursorTarget } from "#app/enums/shop-cursor-target";
```

What a plague this is

File: src/ui/modifier-select-ui-handler.ts [hunk 5]
```diff
@@ -249,11 +250,22 @@ export default class ModifierSelectUiHandler extends AwaitableUiHandler {
...
+      };
 
-      handleTutorial(this.scene, Tutorial.Select_Item).then(() => {
-        this.setCursor(0);
+      setCursorTarget();
+
+      handleTutorial(this.scene, Tutorial.Select_Item).then((res) => {
+        if (res) {
+          setCursorTarget();
+        }
```
Comment (flx-sta) [inline src/ui/modifier-select-ui-handler.ts:268]: I personally think using `set` in the name for a non setter function is confusing.
I'd recommend naming it `refreshCursorTarget()` or `updateCursorTarget()`

File: src/ui/modifier-select-ui-handler.ts [hunk 6]
```diff
@@ -12,6 +12,7 @@ import { allMoves } from "../data/move";
 import * as Utils from "./../utils";
 import Overrides from "#app/overrides";
 import i18next from "i18next";
+import { ShopCursorTarget } from "#app/enums/shop-cursor-target.js";
```
Comment (flx-sta) [inline src/ui/modifier-select-ui-handler.ts:15]:
```suggestion
import { ShopCursorTarget } from "#app/enums/shop-cursor-target";
```

💀

**Review by flx-sta — CHANGES_REQUESTED** [review]

**Review by flx-sta — APPROVED** [review]
Code 🚀 
play-test 🚀 

<sub>BUT: I can already see the bug-reports coming in "THE CURSOR IS GONE!" as `check-team` is so far in the corner. Not your fault though</sub>

**Comment by mKroogz** [comment]
> "THE CURSOR IS GONE!"

😱 😱 😱

File: src/locales/fr/settings.ts [hunk 7]
```diff
@@ -98,5 +98,10 @@ export const settings: SimpleTranslationEntries = {
   "gamepadSupport": "Gamepad Support",
   "showBgmBar": "Titre de la musique",
   "moveTouchControls": "Déplacer les contrôles tactiles",
-  "shopOverlayOpacity": "Opacité boutique"
+  "shopOverlayOpacity": "Opacité boutique",
+  "shopCursorTarget": "Shop Cursor Target",
+  "items": "Items",
+  "reroll": "Reroll",
+  "shop": "Shop",
+  "checkTeam": "Check Team"
```
Comment (Adri1) [inline src/locales/fr/settings.ts:106]:
```suggestion
  "shopCursorTarget": "Choix après relance",
  "items": "Obj. gratuits",
  "reroll": "Relance",
  "shop": "Boutique",
  "checkTeam": "Équipe"
```

File: src/locales/it/settings.ts [hunk 8]
```diff
@@ -98,5 +98,10 @@ export const settings: SimpleTranslationEntries = {
   "gamepadSupport": "Supporto Gamepad",
   "showBgmBar": "Mostra Nomi Musica",
   "moveTouchControls": "Move Touch Controls",
-  "shopOverlayOpacity": "Opacità Finestra Negozio"
+  "shopOverlayOpacity": "Opacità Finestra Negozio",
+  "shopCursorTarget": "Shop Cursor Target",
+  "items": "Items",
+  "reroll": "Reroll",
+  "shop": "Shop",
+  "checkTeam": "Check Team"
```
Comment (NicusPulcis) [inline src/locales/it/settings.ts:106]:
```suggestion
  "shopCursorTarget": "Target Cursore Negozio",
  "items": "Oggetti",
  "reroll": "Rerolla",
  "shop": "Negozio",
  "checkTeam": "Controlla squadra"
```

**Review by NicusPulcis — APPROVED** [review]
italian localized :)

File: src/locales/pt_BR/settings.ts [hunk 9]
```diff
@@ -98,5 +98,10 @@ export const settings: SimpleTranslationEntries = {
   "gamepadSupport": "Suporte para Controle",
   "showBgmBar": "Exibir Nomes das Músicas",
   "moveTouchControls": "Move Touch Controls",
-  "shopOverlayOpacity": "Opacidade da Loja"
+  "shopOverlayOpacity": "Opacidade da Loja",
+  "shopCursorTarget": "Shop Cursor Target",
+  "items": "Items",
+  "reroll": "Reroll",
+  "shop": "Shop",
+  "checkTeam": "Check Team"
```
Comment (josericardo-fo) [inline src/locales/pt_BR/settings.ts:106]:
```suggestion
  "shopCursorTarget": "Alvo do Cursor da Loja",
  "items": "Itens",
  "reroll": "Atualizar",
  "shop": "Loja",
  "checkTeam": "Checar Time"
```

File: src/locales/it/settings.ts (hunk 8 above)
Comment (torranx) [reply src/locales/it/settings.ts:106]: issue with overlapping text
Screenshot 2024-08-21 at 5 13 53 AM

File: src/locales/ko/settings.ts [hunk 10]
```diff
@@ -98,5 +98,10 @@ export const settings: SimpleTranslationEntries = {
   "gamepadSupport": "게임패드 지원",
   "showBgmBar": "BGM 제목 보여주기",
   "moveTouchControls": "터치 컨트롤 이동",
-  "shopOverlayOpacity": "상점 오버레이 투명도"
+  "shopOverlayOpacity": "상점 오버레이 투명도",
+  "shopCursorTarget": "Shop Cursor Target",
+  "items": "Items",
+  "reroll": "Reroll",
+  "shop": "Shop",
+  "checkTeam": "Check Team"
```
Comment (sodaMelon) [inline src/locales/ko/settings.ts:106]:
```suggestion
  "shopCursorTarget": "상점 커서 위치",
  "items": "아이템",
  "reroll": "갱신",
  "shop": "상점",
  "checkTeam": "Check Team"
```

**Review by sodaMelon — COMMENTED** [review]
added korean localization with "cursor"

File: src/locales/zh_CN/settings.ts [hunk 11]
```diff
@@ -98,5 +98,10 @@ export const settings: SimpleTranslationEntries = {
   "gamepadSupport": "手柄支持",
   "showBgmBar": "显示音乐名称",
   "moveTouchControls": "移动触摸控制",
-  "shopOverlayOpacity": "商店显示不透明度"
+  "shopOverlayOpacity": "商店显示不透明度",
+  "shopCursorTarget": "Shop Cursor Target",
+  "items": "Items",
+  "reroll": "Reroll",
+  "shop": "Shop",
+  "checkTeam": "Check Team"
```
Comment (40chyan) [inline src/locales/zh_CN/settings.ts:106]:
```suggestion
  "shopCursorTarget": "商店指针位置",
  "items": "道具",
  "reroll": "刷新",
  "shop": "购买",
  "checkTeam": "检查队伍"
```

**Review by 40chyan — COMMENTED** [review]
zhcn

File: src/locales/it/settings.ts (hunk 8 above)
Comment (NicusPulcis) [reply src/locales/it/settings.ts:106]: thanks for letting me know @torranx , i'll fix it rn

Comment (NicusPulcis) [inline src/locales/it/settings.ts:106]: Is it better like this @torranx ?
```suggestion
  "shopCursorTarget": "Target Cursore Negozio",
  "items": "Oggetti",
  "reroll": "Rerolla",
  "shop": "Negozio",
  "checkTeam": "Squadra"
```

File: src/locales/es/settings.ts [hunk 12]
```diff
@@ -98,5 +98,10 @@ export const settings: SimpleTranslationEntries = {
   "gamepadSupport": "Gamepad Support",
   "showBgmBar": "Show Music Names",
   "moveTouchControls": "Move Touch Controls",
-  "shopOverlayOpacity": "Opacidad de la fase de compra"
+  "shopOverlayOpacity": "Opacidad de la fase de compra",
+  "shopCursorTarget": "Shop Cursor Target",
+  "items": "Items",
+  "reroll": "Reroll",
+  "shop": "Shop",
+  "checkTeam": "Check Team"
```
[the remark this answers is not in the document]
Comment (torranx) [reply src/locales/es/settings.ts:106]: Overlapping text
Screenshot 2024-08-22 at 1 51 15 AM

**Comment by frutescens** [comment]
Tested and it looks good. If it's alright for me to ask, why not have the cursor of modifier-select-ui-hander remember its last held position?

**Comment by torranx** [comment]
> Tested and it looks good. If it's alright for me to ask, why not have the cursor of modifier-select-ui-hander remember its last held position?

I've just tested this rn and if I;m not mistaken, this only matters if the shop is rerolled yes? 
- [One user commented](https://discord.com/channels/1125469663833370665/1235481427924095016/1248315812985438329) that you can accidentally reroll with this
- which leads to [this suggestion](https://discord.com/channels/1125469663833370665/1235481427924095016/1249352981015695500) that I agree with

I think this will be a good starting point for this feature and we could adjust it eventually from user feedback.

File: src/battle-scene.ts [hunk 13]
```diff
@@ -127,6 +128,7 @@ export default class BattleScene extends SceneBase {
   public gameSpeed: integer = 1;
   public damageNumbersMode: integer = 0;
   public reroll: boolean = false;
+  public shopCursorTarget: number = ShopCursorTarget.CHECK_TEAM;
```
Comment (DayKev) [inline src/battle-scene.ts:131]: Shouldn't the default be how it works now?

Comment (torranx) [reply src/battle-scene.ts:131]: It's been discussed [here](https://discord.com/channels/1125469663833370665/1273029080828084234/1275479769596821598)

Comment (DayKev) [reply src/battle-scene.ts:131]: Makes sense. 👍

**Review by DayKev — APPROVED** [review]
Tested it and it works, seems good.

**Comment by frutescens** [comment]
That makes sense, and it looks and works good to me. But, because users can no longer access settings from the shop, they may run into issues when moving from one run to another. Wouldn't it be more straightforward to add a delay before processing inputs?

**Comment by torranx** [comment]
> That makes sense, and it looks and works good to me. But, because users can no longer access settings from the shop, they may run into issues when moving from one run to another. Wouldn't it be more straightforward to add a delay before processing inputs?

I can open settings just fine (had to rebase to beta in case I was missing something):


https://github.com/user-attachments/assets/b5a15c20-bd54-4606-a400-ff426048dfeb

**Comment by torranx** [comment]
F 
branch is still broken

**Comment by frutescens** [comment]
Oh, I'm dumb. Users can no longer access the egg gachas. Sorry, I'm going to drink my caffeine and send an approval in once you're done cleaning up.

**Comment by torranx** [comment]
PR FIXED!

**Review by KimJeongSun — APPROVED** [review]
LGTM! ( and I love it )

**Review by frutescens — APPROVED** [review]

**Pull request merged**